libraries/AP_HAL_Linux/Util_RPI.cpp

```diff
@@ -28,9 +28,34 @@ UtilRPI::UtilRPI()
     _check_rpi_version();
 }
 
+#define MAX_SIZE_LINE 50
 int UtilRPI::_check_rpi_version()
 {
     int hw;
+
+    char buffer[MAX_SIZE_LINE] = { 0 };
+    FILE* f = fopen("/sys/firmware/devicetree/base/model", "r");
```

Comment:
I think this should return the same value as from /proc/cpuinfo. In what case is it different?

Reply:
It is always different, at least on the RPi's I tested it with. /sys/firmware/devicetree/base/model contains a string that looks something like "Raspberry Pi 2 Model B Rev 1.1". My code is reading that "2" (or 3, or nothing in case of RPi 1) after "Raspberry Pi". In /proc/cpuinfo the old method is to detect the chipset ("Hardware: BCM2708") and make the decision based on that. In kernel 4.9 however that line says "Hardware: BCM2835" for all versions of RPi, so there is no way to distinguish between RPi 1 and later versions based on that.

Another method I was considering was based on the "Revision" line in /proc/cpuinfo, which is different for every hardware configuration (even RPi 1 alone has like 10 different revisions), but that would require me to change code in ardupilot/libraries/AP_HAL_Linux/Util.cpp that affects boards other than RPi (namely Bebop[2] and Disco). And since there is a separate Util_RPI.cpp file, I thought it would be better not to put RPi specific code into Util.cpp.